pkg/sql/colexec/COLEXEC.bzl

```diff
@@ -1,44 +1,119 @@
-        )
-        output_files.append(generated_file)
+# Define a file group for all the .eg.go targets.
+def eg_go_filegroup(name):
+    native.filegroup(
+        name = name,
+        srcs = [':{}'.format(rule_name_for(target)) for target, _ in targets],
+    )
+
+# Define gen rules for individual eg.go files.
+def gen_eg_go_rules():
+    # Define some aliases for ease of use.
+    native.alias(
+        name = "execgen",
+        actual = "//pkg/sql/colexec/execgen/cmd/execgen",
+    )
+    native.alias(
+        name = "goimports",
+        actual = "//vendor/github.com/cockroachdb/gostdlib/x/tools/cmd/goimports",
+    )
 
-    # Construct the full set of output files for bazel to be able to analyse.
-    # See https://docs.bazel.build/versions/master/skylark/lib/DefaultInfo.html.
-    return [DefaultInfo(files = depset(output_files))]
+    for target, template in targets:
+        name = rule_name_for(target)
+
+        native.genrule(
+            name = name,
+            srcs = [template],
+            outs = [target],
+            # `$@` lets us substitute in the output path[1]. The symlink below
+            # is frowned upon for genrules[2]. That said, when testing
+            # pkg/sql/colexec through bazel it expects to find the template
+            # files in a path other than what SRCS would suggest. We haven't
+            # really investigated why. For now lets just symlink the relevant
+            # files into the "right" path within the bazel sandbox[3].
+            #
+            # [1]: https://docs.bazel.build/versions/3.7.0/be/general.html#genrule_args
+            # [2]: https://docs.bazel.build/versions/3.7.0/be/general.html#general-advice
+            # [3]: https://github.com/cockroachdb/cockroach/pull/57027
+            cmd = """
+              ln -s external/cockroach/pkg pkg
+              $(location :execgen) -template $(SRCS) \
+                  -fmt=false pkg/sql/colexec/$@ > $@
+              $(location :goimports) -w $@
+              """,
+            tools = [":execgen", ":goimports"],
+        )
 
-generate = rule(
-    implementation = _generate_impl,
-    # Source all the necessary template files, output targets, the execgen and goimports binaries.
-    attrs = {
-        "templates": attr.label_list(mandatory = True, allow_files = ["tmpl.go"]),
-        "targets": attr.string_list(mandatory = True),
-        "execgen": attr.label(mandatory = True, executable = True, allow_files = True, cfg = "exec"),
-        "goimports": attr.label(mandatory = True, executable = True, allow_files = True, cfg = "exec"),
-    },
-)
+# TODO(irfansharif): We should be able to use `gen_eg_go_rules` to generate
+# like_ops.eg.go. It's special-cased here because execgen reads two separate
+# template files[1] when generating like_ops.eg.go.
+#
+# [1]: https://github.com/cockroachdb/cockroach/blob/1f23ef2b4e/pkg/sql/colexec/execgen/cmd/execgen/like_ops_gen.go#L48
+def gen_like_ops_rule(name, templates, target):
```

Comment:
re-reading this, can we just make the targets be something like
```

targets = [
    ...,
    ('is_null_ops.eg.go', 'is_null_ops_tmpl.go'),
    ('like_ops.eg.go', '"selection_ops_tmpl.go", ["proj_const_ops_tmpl.go"])
    ...
] # feel free to make this a class

for name, template, extra_rules in ...:
    native.genrule(
            name = name,
            srcs = [template] + extra_rules,
            outs = [target],
            # `$@` lets us substitute in the output path[1]. The symlink below
            # is frowned upon for genrules[2]. That said, when testing
            # pkg/sql/colexec through bazel it expects to find the template
            # files in a path other than what SRCS would suggest. We haven't
            # really investigated why. For now lets just symlink the relevant
            # files into the "right" path within the bazel sandbox[3].
            #
            # [1]: https://docs.bazel.build/versions/3.7.0/be/general.html#genrule_args
            # [2]: https://docs.bazel.build/versions/3.7.0/be/general.html#general-advice
            # [3]: https://github.com/cockroachdb/cockroach/pull/57027
            cmd = """
              ln -s external/cockroach/pkg pkg
              $(location :execgen) -template $(SRCS) \
                  -fmt=false pkg/sql/colexec/$@ > $@
              $(location :goimports) -w $@
              """,
            tools = [":execgen", ":goimports"],
        )

Reply:
Meh, I had that originally. I want to explicitly special case like_ops.eg.go away cause the execgen code, as written, is doing something non-obvious.

```
execgen -template <doesn't list out everything needed> pkg/sql/colexec/like_ops.eg.go
```

Should probably fixed there and then the special case would go away.

Reply:
is it going to go away? if not i'd rather have ^. if so, maybe tie an issue to it.
i'll let you judge here, LGTM otherwise